### PR TITLE
[Desktop] Fix changing WindowPlacement from Fullscreen to Maximized

### LINKED ISSUE #4380: [Desktop] Closing full screen is not working correctly
**Describe the bug**
When I share the window state placement to `WindowPlacement.FullScreen` everything works fine but I change it back to `WindowPlacement.Maximized` it remains full screen.

**Affected platforms**
- Desktop

**Versions**
- Kotlin version*: 1.9.22
- Compose Multiplatform version*: 1.6.0-rc02
- OS version(s)* (required for Desktop and iOS issues): MacOs 13.0.1
- OS architecture (x86 or arm64): arm64
- JDK (for desktop issues): 17

**To Reproduce**
Steps and/or the code snippet to reproduce the behavior:
1. Change window state placement to `WindowPlacement.FullScreen`
2. Change window state placement to `WindowPlacement.Maximized`
3. You will see that the window is still in full-screen mode.

This is a screen recording describing the issue:
https://github.com/JetBrains/compose-multiplatform/assets/41842296/4d7df284-b367-4452-bc34-8e7965425597

**Expected behavior**
The full-screen mode should be disabled.

This is a screen recording describing the expected behavior:
https://github.com/JetBrains/compose-multiplatform/assets/41842296/0b42c46e-b761-4f5b-b8d0-9e06e1ee366e

**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by MohamedRejeb**
## Proposed Changes

When the `WindowPlacement` is set to `Maximized`, I changed the `isFullscreen` value to false to make sure that we exit fullscreen.
It's fixing the issue in mac, I still didn't try it on windows to make sure that it's not causing any issues.

## Testing

Test: Describe how you tested your changes. Note that this line (with `Test:`) is required, your PR will not build without it!

I wasn't able to test this change since `isFullscreen` is private.

## Issues Fixed

Fixes: https://github.com/JetBrains/compose-multiplatform/issues/4380

## Google CLA
You need to sign the Google Contributor’s License Agreement at https://cla.developers.google.com/.
This is needed since we synchronise most of the code with Google’s AOSP repository. Signing this agreement allows us to synchronise code from your Pull Requests as well.

### CODE REVIEW

**Comment by MohamedRejeb** [comment]
This fixes the issue, but it looks like it causes a different issue.
Now if you set window placement to floating -> fullscreen -> maximized, the placement is going to be changed to floating instead of maximized. It looks like the `isFullscreen = false` is going to return the window to the latest state before opening fullscreen.
I will add this condition to the test.
Any ideas how to fix this edge case?

**Comment by MohamedRejeb** [comment]
I was able to use this workaround, the first set placement to Maximized will exit the fullscreen, then I wait for the full screen to be exited and then I set the placement to Maximized again.

```kotlin
scope.launch {
    window.placement = WindowPlacement.Maximized
    while (window.placement == WindowPlacement.Fullscreen) {
        withFrameMillis {  }
    }
    window.placement = WindowPlacement.Maximized
}
```

So for this to work correctly, we need to add a check when the window is fullscreen and you change the placement to Maximized exit the fullscreen first and wait for it to be exited successfully then maximize the window.

**Comment by MohamedRejeb** [comment]
Using something like this can fix the issue, but I'm not sure if you are ok with using runBlocking here. We need to wait for fullscreen to exit before changing the `isMaximized` value.
There's also another possible solution by relying on the listeners in `Window.desktop.kt` and correct the `isMaximized` value after exiting fullscreen.

Screenshot 2024-02-28 at 5 56 45 PM

File: compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/window/WindowStateTest.kt [hunk 1]
```diff
@@ -310,6 +310,130 @@ class WindowStateTest {
...
+                window = this.window
+            }
+        }
+
+        awaitIdle()
+
+        // Enter fullscreen from maximized
+        state.placement = WindowPlacement.Maximized
+        awaitIdle()
+        assertThat(window.placement).isEqualTo(WindowPlacement.Maximized)
```
Comment (igordmn) [inline compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/window/WindowStateTest.kt:394]: Tests [fail](https://teamcity.jetbrains.com/buildConfiguration/JetBrainsPublicProjects_Compose_DesktopTestsForPullRequests/4508579?buildTab=overview&expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&expandBuildTestsSection=true&hideProblemsFromDependencies=false) on Linux in these lines:
```
expected: Maximized
but was : Floating
```

**Review by igordmn — COMMENTED** [review]
Thanks!

>  I still didn't try it on windows

It seems work on Windows, but it fails on Linux. See the failed tests.

> It's fixing the issue in mac

The mac fullscreen logic is overriden in [Skiko](https://github.com/JetBrains/skiko/blob/2f711b773d953115be58bad12276658366a81674/skiko/src/awtMain/objectiveC/macos/Drawlayer.mm#L233).

First, it violates the contract: setting `isFullscreen = false` should immediately cause `isFullscreen = false`. When this contract is fixed, your workaround won't work.

Second, it performs setting fullscreen asynchronously calling `performSelectorOnMainThread`. that is why `isMaximized = true` didn't work.

We probably can't wait for AppKit main thread from Swing thread, as it can cause deadlocks (main thread can wait for Swing thread as well).

I don't have an easy solution in mind

Maybe to fix the mac issue we need to provide `skialayer.placement`, so it can schedule `fullscreen = false` and `maximized = true` on the main AppKit thread at once.

To fix the Linux issue, we need to experiment and look at the Swing sources. If Swing itself doesn't support this use case, we have to write a native code instead.

**Comment by kropp** [comment]
I'm closing this PR for now as it has unresolved reviewer comments and conflicts with the main branch.
